### PR TITLE
fix aggregation (avg and quantiles) over empty decimal columns.

### PR DESCRIPTION
**Pull request by akonyaev90**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Bug Fix

Short description (up to few sentences):
fix aggregation (avg and quantiles) over empty decimal columns.

```
CREATE TABLE test
(
    `x` Int32, 
    `d` Decimal(10, 3)
)
ENGINE = MergeTree()
ORDER BY x;

SELECT avg(d)
FROM test

↑ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) Received exception from server (version 19.14.3):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: AggregateFunctionAvg with zero values. 

SELECT quantiles(0.1, 0.5, 0.9)(d)
FROM test

┌─quantiles(0.1, 0.5, 0.9)(d)────────────────────────────────────────────┐
│ [--9223372036854775.808,--9223372036854775.808,--9223372036854775.808] │
└────────────────────────────────────────────────────────────────────────┘


```

With this fix:
```

SELECT avg(d)
FROM test

┌─avg(d)─┐
│  0.000 │
└────────┘

SELECT quantiles(0.1, 0.5, 0.9)(d)
FROM test

┌─quantiles(0.1, 0.5, 0.9)(d)─┐
│ [0.000,0.000,0.000]         │
└─────────────────────────────┘


```

Tests added.

### CODE REVIEW

File: dbms/src/AggregateFunctions/ReservoirSampler.h [hunk 1]
```diff
@@ -108,9 +108,12 @@ class ReservoirSampler
       */
     double quantileInterpolated(double level)
     {
-        if (samples.empty())
+        if (samples.empty()) {
+            if (DB::IsDecimalNumber<T>) {
```
Comment (4ertus2) [inline dbms/src/AggregateFunctions/ReservoirSampler.h]: the same: no braces here.

File: dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+SELECT uniqUpTo(10)(a), uniqUpTo(10)(b), uniqUpTo(10)(c) FROM decimal WHERE a >= 0 AND a < 5;
+SELECT uniqUpTo(10)(a), uniqUpTo(10)(b), uniqUpTo(10)(c) FROM decimal WHERE a >= 0 AND a < 10;
+
+SELECT argMin(a, b), argMin(a, c), argMin(b, a), argMin(b, c), argMin(c, a), argMin(c, b) FROM decimal;
+SELECT argMin(a, b), argMin(a, c), argMin(b, a), argMin(b, c), argMin(c, a), argMin(c, b) FROM decimal WHERE a > 0;
+SELECT argMax(a, b), argMax(a, c), argMax(b, a), argMax(b, c), argMax(c, a), argMax(c, b) FROM decimal;
+SELECT argMax(a, b), argMax(a, c), argMax(b, a), argMax(b, c), argMax(c, a), argMax(c, b) FROM decimal WHERE a < 0;
+
+SELECT median(a), median(b), median(c) as x, toTypeName(x) FROM decimal;
```
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql]: We discussed that toTypeName(a) and toTypeName(b) should return Decimal(9, 4) and Decimal(18, 8)

File: dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+SELECT count(a), count(b), count(c) FROM decimal;
+SELECT [min(a), max(a)], [min(b), max(b)], [min(c), max(c)] FROM decimal;
+
+SELECT sum(a), sum(b), sum(c), sumWithOverflow(a), sumWithOverflow(b), sumWithOverflow(c) FROM decimal;
+SELECT sum(a), sum(b), sum(c), sumWithOverflow(a), sumWithOverflow(b), sumWithOverflow(c) FROM decimal WHERE a > 0;
+SELECT sum(a), sum(b), sum(c), sumWithOverflow(a), sumWithOverflow(b), sumWithOverflow(c) FROM decimal WHERE a < 0;
+SELECT sum(a+1), sum(b+1), sum(c+1), sumWithOverflow(a+1), sumWithOverflow(b+1), sumWithOverflow(c+1) FROM decimal;
+SELECT sum(a-1), sum(b-1), sum(c-1), sumWithOverflow(a-1), sumWithOverflow(b-1), sumWithOverflow(c-1) FROM decimal;
+
+SELECT avg(a), avg(b), avg(c) FROM decimal;
```
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql]: Need also toTypeName(a), toTypeName(b), toTypeName(c) columns.

**Review by 4ertus2 — CHANGES_REQUESTED** [review]

File: dbms/src/AggregateFunctions/ReservoirSampler.h (hunk 1 above)
Comment (akonyaev90) [reply dbms/src/AggregateFunctions/ReservoirSampler.h]: code style fixed
I started using check-style utility

File: dbms/src/AggregateFunctions/ReservoirSampler.h [hunk 4]
```diff
@@ -108,9 +108,12 @@ class ReservoirSampler
       */
     double quantileInterpolated(double level)
     {
-        if (samples.empty())
+        if (samples.empty()) {
```
[the remark this answers is not in the document]
Comment (akonyaev90) [reply dbms/src/AggregateFunctions/ReservoirSampler.h]: code style fixed

File: dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql (hunk 3 above)
Comment (akonyaev90) [reply dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql]:
```
SELECT 
    quantiles(0.5, 0.6, 0.99999)(d) AS q, 
    toTypeName(q), 
    avg(d) AS a, 
    toTypeName(a)
FROM test2

┌─q───────────────────┬─toTypeName(quantiles(0.5, 0.6, 0.99999)(d))─┬─────a─┬─toTypeName(avg(d))─┐
│ [3.500,3.800,4.999] │ Array(Decimal(10, 3))                       │ 3.500 │ Decimal(10, 3)     │
└─────────────────────┴─────────────────────────────────────────────┴───────┴────────────────────┘

1 rows in set. Elapsed: 0.017 sec. 

```
result types for avg fixed.
tests updated

File: dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql (hunk 2 above)
Comment (akonyaev90) [reply dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql]: I changed types in testing table for empty test.
now it like this:
```
    a Decimal(6, 4),
    b Decimal(16, 7),
    c Decimal(20, 8)
```

Comment (akonyaev90) [reply dbms/tests/queries/0_stateless/00700_decimal_empty_aggregates.sql]: for quantile and median added test for result types

File: dbms/src/AggregateFunctions/AggregateFunctionAvg.h [hunk 5]
```diff
@@ -43,29 +43,31 @@ template <typename T, typename Data>
...
         : IAggregateFunctionDataHelper<Data, AggregateFunctionAvg<T, Data>>(argument_types_, {})
         , scale(0)
+        , precision(0)
     {}
 
     /// ctor for Decimals
     AggregateFunctionAvg(const IDataType & data_type, const DataTypes & argument_types_)
         : IAggregateFunctionDataHelper<Data, AggregateFunctionAvg<T, Data>>(argument_types_, {})
         , scale(getDecimalScale(data_type))
+        , precision(getDecimalPrecision(data_type))
```
Comment (4ertus2) [inline dbms/src/AggregateFunctions/AggregateFunctionAvg.h]: We'd better save old logic here cause it's inconsistent with other places. We have no logic to check real decimal overflows so precision is something misleading. The only one place we check it is String to Decimal conversion.

File: dbms/src/DataTypes/DataTypesDecimal.h [hunk 6]
```diff
@@ -243,6 +243,17 @@ inline UInt32 getDecimalScale(const IDataType & data_type, UInt32 default_value
     return default_value;
 }
 
+inline UInt32 getDecimalPrecision(const IDataType & data_type, UInt32 default_value = std::numeric_limits<UInt32>::max())
```
Comment (4ertus2) [inline dbms/src/DataTypes/DataTypesDecimal.h]: Misleading function

**Review by 4ertus2 — CHANGES_REQUESTED** [review]

File: dbms/src/AggregateFunctions/AggregateFunctionAvg.h (hunk 5 above)
Comment (akonyaev90) [reply dbms/src/AggregateFunctions/AggregateFunctionAvg.h]: fixed

```
CREATE TABLE test_23
(
    `d1` Decimal(6, 2), 
    `d2` Decimal(11, 4), 
    `d3` Decimal(28, 1)
)
ENGINE = Memory

Ok.

0 rows in set. Elapsed: 0.012 sec. 

akonyaev-laptop :) select toTypeName(avg(d1)), toTypeName(avg(d2)),toTypeName(avg(d3)) from test_23

SELECT 
    toTypeName(avg(d1)), 
    toTypeName(avg(d2)), 
    toTypeName(avg(d3))
FROM test_23

┌─toTypeName(avg(d1))─┬─toTypeName(avg(d2))─┬─toTypeName(avg(d3))─┐
│ Decimal(9, 2)       │ Decimal(18, 4)      │ Decimal(38, 1)      │
└─────────────────────┴─────────────────────┴─────────────────────┘

```

tests fixed

File: dbms/src/DataTypes/DataTypesDecimal.h (hunk 6 above)
Comment (akonyaev90) [reply dbms/src/DataTypes/DataTypesDecimal.h]: removed.

**Review by 4ertus2 — APPROVED** [review]
OK

**Pull request merged**